notificator/notificator_queue.py

```diff
@@ -0,0 +1,221 @@
+        self.db_pool = pool
+        self.notified_accounts = None
+        self.period = period
+
+    async def init(self):
+        """Async constructor"""
+        self.notified_accounts = await self._create_notif_accs_map()
+
+    async def stop(self):
+        """Stops queue loop"""
+        await self.notifications_topic.stop()
+
+    async def _create_notif_accs_map(self):
+        """Creates a map with already notified customers with notification event types"""
+        res = {}
+        async with self.db_pool.acquire() as conn:
+            async with conn.transaction():
+                cur = conn.cursor("""SELECT rh_account_id, cve_id, notif_type FROM notified_accounts""")
+                async for notif_row in cur:
+                    rh_acc = notif_row[0]
+                    cve_id = notif_row[1]
+                    notif_type = NotificationType(notif_row[2])
+                    if (rh_acc, cve_id) not in res:
+                        res[(rh_acc, cve_id)] = set()
+                    res[(rh_acc, cve_id)].add(notif_type)
+        return res
+
+    @staticmethod
+    def _create_notif_context(cve):
+        """Create dict of context values for templated emails"""
+        return {
+            "reported_cve": cve,
+        }
+
+    def _send_kafka_notif(self, account_id: str, event_type: str, event_context: dict, only_admins=False, ignore_user_preferences=False, loop=None):
+        """
+        Sends kafka msg to notification kafka.
+
+        account_id: insights id of user
+        event_type: Notifications event type for vulnerability
+        event_context: Dict with data for email templates
+        only_admins: If notif should be sent only to orgs admin
+        ignore_user_preferences: Ignores users settings in notifications app
+        """
+        msg = {
+            "bundle": "rhel",
+            "application": "vulnerability",
+            "version": "v1.1.0",
+            "account_id": account_id,
+            "event_type": event_type,
+            "context": json.dumps(event_context),  # context json needs to be escaped
+            "events": [],
+            "recipients": [
+                {
+                    "only_admins": only_admins,
+                    "ignore_user_preferences": ignore_user_preferences,
+                }
+            ],
+        }
+        LOGGER.debug("Sending notification: %s", msg)
+        self.notifications_topic.send(msg, loop=loop)
+
+    async def _register_notified_acc(self, cve_id: int, notif_events: [NotificationType], rh_account_id):
+        """Registers new notified accounts into notified_accounts table"""
+        if not notif_events:
+            return
+
+        tuples = [(rh_account_id, cve_id, notif_event.value) for notif_event in notif_events]
+        async with self.db_pool.acquire() as conn:
+            await conn.executemany("""INSERT INTO notified_accounts (rh_account_id, cve_id, notif_type) VALUES ($1, $2, $3)""", tuples)
```

Comment:
Better handle a conflict.

Reply:
This conflict should not happen, I would leave it there if some bug would appear

Reply:
If it's running as singleton, then no. When the service will run in multiple instances, then it can happen easily. And there is a bit pressure on us from app-sre to make all service scalable. Also I'm afraid in this case it'd send some duplicate messages too, as each container is maintaining its own `self.notified_accounts` cache.

Reply:
So I have changed the `notified_accounts` cache, if the notification type is not in the cache, it gets fetched for given `cve_id` and `rh_account_id` pair and stored.